### ISSUE TITLE
Allow users to set the origin of Control Nodes

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#129]
**Describe the project you are working on:**
Card game, Action game with items, menus.

**Describe the problem or limitation you are having in your project:**
When working on menus, anything with drag'n'drop functionality, items, cards, or even a simple debug text label, I find myself in dire need to have control over the Control nodes origin every single time. 
I work with a Control nodes as well as those inheriting from Control, but Godot currently does not seem to offer a way to adjust the origin. `rect_pivot_offset` is not an option since it is an absolute value (pixels I assume), not a percentage.
The default origin for all Control nodes is the top right corner. This makes centering and scaling and positioning incredibly tedious and intuitive.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
Controling scale, positioning and grow direction is currently a nightmare from UX perspective. It forces you to use multiple nested nodes and properties of those nodes and absolut pixel values.
This proposal would reduce all that to just one single setting.

Personally I would think having the default origin of Control nodes at the center would make the most sense. As this would allow Control nodes like ColorRect, Lable, Button, Control to behave like Sprite, CollisionObject2D, Polygon2D, Light2D, Ysort, Node2D when they are spawned.
But I understand that would massively break compatibility, so as second best option I would really like to be allowed to change it at least.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![control_origin](https://user-images.githubusercontent.com/47016402/66269758-d3cb8600-e84c-11e9-96d6-bb519ccf2de2.png)

**Describe implementation detail for your proposal (in code), if possible:**
I don't know any C++

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It's the first thing I would want to adjust every single time I work with any kind Node that inherits from Control, so yes, it would be used often.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It's mainly a (massive) UI/UX improvement of existing features.

**Comment by hammeron-art** [comment]
That's what Pivot Offset is for. The main issue is that it doesn't update automatically. It should work like size and margin which update when related properties change.

Predefined origins like the screenshot will add another limitation. It must be values. I do agree it could be percentages like the anchors but there is a way of implementing it without breaking compatibility. Keep Pivot Offset as it is and add _Pivot Modes_ such as _Keep Position_ (current behavior) and _Proportional Scale_ (update pivot when related properties change).

**Comment by golddotasksquestions** [comment]
@hammeron-art 
No, my main issue is the user experience. Making Pivot Offset change automatically might help with the negative symptoms short term, but the inherently bad design of the user experience still prevails. No halve way decent software with layout capabilities has such a crazy backwards workflow and at the core of this all is the fixed origin position.

If the user does not want the origin to be in the top left corner, the user should be able to set it somewhere else.

**Comment by hammeron-art** [comment]
I'm not understanding your reasoning. The origin isn't the left corner but the pivot offset. If you want the origin at the center, set the pivot offset to the rect size*0.5. You would have to set the pivot once and the automatic update would keep it at the center when the rect size changes. It's the same effect as using percentage.

**Comment by golddotasksquestions** [comment]
Imagine a centered Label you cast some text to. Every time the text length changes, the Pivot Offest value will have to automatically change as well. Do you think this is good design?

**Comment by hammeron-art** [comment]
I agreed from the start that percentage for UI is better. My suggestion of auto update is just a way of keeping compatibility or if they are not interested in changing the current design. The size and margin already work like this so it makes sense to make the pivot works like this as well.

**Comment by lentsius-bark** [comment]
I'd like to chip in to say that even if the proposed solution is not the best approach, I do hope the pivot editing options gets an update for UI. I for one remember playing with horizontal and vertical growth and pivot position to achieve desired UI functionality which took me a lot of testing and fiddling, which done with percentages would have been a single click away.

**Comment by vitorbalbio** [comment]
I'm also having problems since there's no way to set the Control origin other than the left superior corner. So i need to hardcode offsets for each component when i do something like project 2D elements on 3D scenes.

**Comment by HybridEidolon** [comment]
I am working on some controls for a game UI using the Control hierarchy. While the container system works great for conventional desktop UIs, the lack of an "origin" makes it difficult to use for Controls that are free-standing in screen space. For example, I want to be able to programmatically place a speech dialog at a specific location on-screen, with its origin being a point between two PanelContainers in a VBoxContainer. This is surprisingly hard to accomplish with the current system, especially if the Control changes size at all (e.g. the text shown resizes its sub-container).

**Comment by kilojool** [comment]
I also would like some way to change the origin of control nodes. If the pivot offset affected position also, then that would be enough for me. Not sure if that would cause other problems though.

**Comment by hawkerm** [comment]
I started playing with Godot a bit and wanted to animate some text rotating about its center-point. This was the first issue I encountered, was hard to find this thread too as I saw a way to easily set the anchor point in the UI, but no way to do something similar for the origin of the control. Finally after like 20 minutes stumbled upon the `pivot_offset` properties in the inspector.

**Comment by kilojool** [comment]
After a couple of years with Godot, I honestly think this is what still gives me the most headaches. ;) There are workarounds, like using a dummy parent node with size zero, or some containers also can do the trick, but I would prefer not to clog up the scene tree too much.
For cards, arrows, and similar objects where the rotation and origin is generally supposed to be the same point, this would hel alot. 
Also simple things like resizing a control node while keeping it centered is really unintuitive if not using CenterContainer. A relative origin offset would likely make that so much easier too.

**Comment by YuriSizov** [comment]
For rotation and scaling we have a PR in works that adds an offset ratio on top of the pixel offset for the pivot center: https://github.com/godotengine/godot/pull/70646

**Comment by gpergrossi** [comment]
From the behavior I'm seeing, it looks like a mistake is being made in the code (which I haven't bothered to locate myself since I've never worked with godot source). Basically, I think the pivot_offset is getting _added_ to the position of the node rather than _subtracted_.

Take for example a label with size (200, 100) with its position at (0, 0). As you expect it rotates around (0,0) and its center is at (100, 50).

Now I start adding a pivot offset because I really want it to rotate around the center. So I set the offset to (100, 50).

The node's position is _still_ (0,0), but now it draws the node with its center at (100, 50) instead of (0,0). It would make all the sense in the world to assume this comes from (0,0) + pivot_offset. But it _should_ be (0, 0) - pivot_offset. So as I move the pivot the node actually centers itself around that pivot.

As others have mentioned the other issue is that it should be a ratio in the UI at least, so I'm in favor of https://github.com/godotengine/godot/pull/70646

**Comment by bitbendergames** [comment]
I too vote for a relative pivot position in the inspector. This seems like very very basic UI functionality that many people need, judging from the amount of upvotes in the OP.

**Comment by TGRRRR** [comment]
I would love to have this feature too!

**Comment by AThousandShips** [comment]
@TGRRRR Please don't bump without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by Meorge** [comment]
Not being able to specify the origin/pivot point for controls (easily?) is one of the big frustrations I've continued to have while attempting to make game UI in Godot. I was directed to issue https://github.com/godotengine/godot/pull/70646 independently of this thread for the concept of pivot offset ratio and have been hoping to contribute to that or, at least, show interest in it and make it a higher-priority item.

gpergossi's explanation of the difficulties with the current system really echo my own thoughts. In Unity, I remember I was able to define any point on a UI item as its origin and pivot point, and then easily position (and animate etc) it based on that custom-set point. I've struggled to do the same in Godot.

Might these issues might be worth trying to combine into more of a general overhaul of Control node positioning? I recall Unity's UI tools in these areas being much more flexible than Godot's, so perhaps we could take some cues from it. I imagine this would have to be a change for a more major release, though, and with tools for translating old Control node setups to whatever new system is devised.

If there are any more concrete ways I could help with getting these improvements to Godot through, I'd love to try! My experience with the engine code itself, as well as layout algorithms like this, is... limited, to say the least, but I suppose that's true for anyone before they start getting experience.

**Comment by Landeplage** [comment]
I too miss this feature, as someone who is coming from Unity and learning Godot. Setting pivot based on pixel coordinates doesn't make sense to me. A ratio from 0 to 1 would be better in most cases, IMO.

Someone who needs to use pixel coordinates could use the current workaround that those who need 0 to 1 use now; calculating it using scripting.

My use case revolves around UI layout and animation; for example how a button icon moves when it is scaled based on a hover state, or a menu show/hide animation.
